File: July2021/C++/CountVowelsPermutation.cpp
class Solution {
public:
    const int m=1e9+7;
    int countVowelPermutation(int n) {
        long long dp[5][n+1];
        long long ans=0;
        dp[0][1]=dp[1][1]=dp[2][1]=dp[3][1]=dp[4][1]=1;
        for(long long j=2; j<=n; j++)
        {
            dp[0][j]=dp[1][j-1];
            dp[1][j]=(dp[0][j-1]+dp[2][j-1])%m;
            dp[2][j]=(dp[0][j-1]+dp[1][j-1]+dp[3][j-1]+dp[4][j-1])%m;
            dp[3][j]=(dp[2][j-1]+dp[4][j-1])%m;
            dp[4][j]=dp[0][j-1];
        }
        for(long long j=0; j<5; j++)
        {
            ans=(ans+dp[j][n])%m;
        }
        return ans;
    }
};

File: July2021/CountVowelsPermutation.java
class Solution {

  private Map<String,Long> map = new HashMap<>();
  private int MOD = 1000000007;

  public int countVowelPermutation(int n) {
    long total = 0;
    char[] charSet = new char[]{'a','e','i','o','u'};
    for(char c : charSet) {
      total = (total + helper(n-1,c)) % MOD;
    }
    return (int)total;
  }

  public long helper(int remaingChar, char prevChar) {
    if(remaingChar == 0) return 1;
    String key = "" + remaingChar + prevChar;
    if(map.containsKey(key)) return map.get(key);
    long total = 0;
    switch(prevChar) {
      case 'a': total = (helper(remaingChar - 1,'e')) % MOD;
        break;
      case 'e': total = (helper(remaingChar - 1,'a') + helper(remaingChar - 1, 'i')) % MOD;
        break;
      case 'i': total = (helper(remaingChar - 1,'a') + helper(remaingChar - 1, 'e') + helper(remaingChar - 1, 'o') + helper(remaingChar - 1, 'u')) % MOD;
        break;
      case 'o': total = (helper(remaingChar - 1,'i') + helper(remaingChar - 1,'u')) % MOD;
        break;
      case 'u': total = (helper(remaingChar - 1,'a')) % MOD;
        break;
    }
    map.put(key , total);
    return total;
  }
}